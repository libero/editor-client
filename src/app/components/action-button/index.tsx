import React from 'react';
import { Button, PropTypes } from '@material-ui/core';
import { mapValues } from 'lodash';
import AddIcon from '@material-ui/icons/Add';

import { useActionButtonStyles } from './styles';

type MuiButtonVariant = 'text' | 'outlined' | 'contained';

const VARIANTS = {
  containedWarning: {
    variant: 'contained' as MuiButtonVariant,
    classes: {
      root: 'containedRedButton'
    }
  },
  outlinedWarning: {
    variant: 'outlined' as MuiButtonVariant,
    classes: {
      root: 'outlinedRedButton'
    }
  },
  primaryContained: {
    variant: 'contained' as MuiButtonVariant,
    color: 'primary',
    classes: {}
  },
  secondaryOutlined: {
    variant: 'outlined',
    classes: {}
  },
  addEntity: {
    size: 'small',
    classes: { root: 'addEntityButton_root', startIcon: 'addEntityButton_icon' },
    variant: 'text',
    startIcon: <AddIcon />
  }
} as {
  [key: string]: {
    variant: MuiButtonVariant;
    classes: Record<string, string>;
    color?: PropTypes.Color;
  };
};

export type VariantType = keyof typeof VARIANTS;

interface ActionButtonProps {
  onClick: () => void;
  title: string;
  variant: VariantType;
  className?: string;
}

export const ActionButton: React.FC<ActionButtonProps> = (props) => {
  const classes = useActionButtonStyles();
  const variantProps = VARIANTS[props.variant];
  const variantClasses = mapValues(variantProps.classes, (className: string) => classes[className]);
  const buttonProps = { ...variantProps, classes: variantClasses };
  return (
    <Button onClick={props.onClick} {...buttonProps} className={props.className}>
      {props.title}
    </Button>
  );
};
