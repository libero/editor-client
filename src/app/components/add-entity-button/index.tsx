import React from 'react';
import AddIcon from '@material-ui/icons/Add';
import { Button } from '@material-ui/core';
import { useAddEntityBtnStyles } from './styles';

export const AddEntityButton: React.FC = () => {
  const classes = useAddEntityBtnStyles();
  return (
    <Button
      size="small"
      classes={{ root: classes.root, startIcon: classes.icon }}
      variant="text"
      startIcon={<AddIcon />}
    >
      Add Author
    </Button>
  );
};
