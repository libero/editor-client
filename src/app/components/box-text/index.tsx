import React from 'react';
import { NodeView, EditorView } from 'prosemirror-view';
import { Node as ProsemirrorNode } from 'prosemirror-model';
import ReactDOM from 'react-dom';
import { ThemeProvider } from '@material-ui/core/styles';
import { NodeSelection } from 'prosemirror-state';

import { BoxTextEditor, BoxTextEditorHandle } from './box-text-editor';
import { theme } from '../../styles/theme';
import { NodeViewContext } from '../../utils/view.utils';

export class BoxTextNodeView implements NodeView {
  dom?: HTMLElement;
  boxTextEditor: React.RefObject<BoxTextEditorHandle>;

  constructor(private node: ProsemirrorNode, private view: EditorView, private getPos: () => number) {
    this.dom = document.createElement('section');
    this.boxTextEditor = React.createRef();
    ReactDOM.render(
      <ThemeProvider theme={theme}>
        <NodeViewContext.Provider
          value={{
            view: this.view,
            getPos: this.getPos,
            getNode: () => this.node
          }}
        >
          <BoxTextEditor ref={this.boxTextEditor} node={this.node} onDelete={this.handleDelete} />
        </NodeViewContext.Provider>
      </ThemeProvider>,
      this.dom
    );

    this.dom.addEventListener('drop', this.handleDrop, true);
  }

  handleDrop = (event: Event): void => {
    event.stopPropagation();
    event.preventDefault();
  };

  handleDelete = (): void => {
    const change = this.view.state.tr
      .setSelection(new NodeSelection(this.view.state.doc.resolve(this.getPos())))
      .deleteSelection();
    this.view.dispatch(change);
  };

  update(node: ProsemirrorNode): boolean {
    this.node = node;
    this.boxTextEditor.current.updateContent(node);
    return true;
  }

  stopEvent(evt): boolean {
    return this.dom.contains(evt.target) || /drag/.test(evt.type);
  }

  destroy(): void {
    this.dom.removeEventListener('drop', this.handleDrop, true);
  }

  ignoreMutation(): boolean {
    return true;
  }
}
