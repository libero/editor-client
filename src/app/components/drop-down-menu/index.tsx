import React, { useCallback, useRef } from 'react';
import { Popper, ClickAwayListener, Paper, MenuList, Button, MenuItem } from '@material-ui/core';
import ArrowDropDownIcon from '@material-ui/icons/ArrowDropDown';

import { useDropDownStyles } from './styles';

export interface DropDownMenuItemProps {
  title: string;
  enabled: boolean;
  selected?: boolean;
  action(): void;
}

export interface DropDownMenuProps {
  title: string;
  entries: DropDownMenuItemProps[];
}

export const DropDownMenu: React.FC<DropDownMenuProps> = ({ title, entries }) => {
  const [open, setOpen] = React.useState(false);
  const menuRef = useRef();
  const handleOpenMenuClick = useCallback(() => setOpen(true), [setOpen]);
  const handleMenuClose = useCallback(() => setOpen(false), [setOpen]);
  const classes = useDropDownStyles();

  const handleMenuItemClick = useCallback((action) => {
    if (action) {
      action();
    }
    setOpen(false);
  }, []);

  const menuItems = entries.map((entry, index) => {
    return (
      <MenuItem
        disabled={!entry.enabled}
        key={index}
        selected={entry.selected}
        onClick={handleMenuItemClick.bind(null, entry.action)}
      >
        {entry.title}
      </MenuItem>
    );
  });

  return (
    <div>
      <Button ref={menuRef} aria-haspopup="true" onClick={handleOpenMenuClick} className={classes.button}>
        {title}
        <ArrowDropDownIcon />
      </Button>
      <Popper anchorEl={menuRef.current} role={undefined} transition disablePortal open={open}>
        <Paper>
          <ClickAwayListener onClickAway={handleMenuClose}>
            <MenuList autoFocusItem={open}>{menuItems}</MenuList>
          </ClickAwayListener>
        </Paper>
      </Popper>
    </div>
  );
};
