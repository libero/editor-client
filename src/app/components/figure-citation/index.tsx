import React from 'react';
import ReactDOM from 'react-dom';
import { Node as ProsemirrorNode } from 'prosemirror-model';
import { EditorView, NodeView } from 'prosemirror-view';
import { ThemeProvider } from '@material-ui/core/styles';

import { theme } from '../../styles/theme';
import { FigureCitationEditorPopup, FiguresListEntry, UNLABELLED_FIGURE_TEXT } from './figure-citation-editor-popup';
import { TextSelection } from 'prosemirror-state';
import { store } from '../../store';
import { Provider } from 'react-redux';

export class FigureCitationNodeView implements NodeView {
  dom?: HTMLAnchorElement;
  contentDOM?: HTMLSpanElement;
  figureEditorContainer: HTMLDivElement;

  constructor(private node: ProsemirrorNode, private view: EditorView, private getPos: () => number) {
    this.dom = document.createElement('a');
    this.dom.style.cursor = 'pointer';
    this.dom.addEventListener('click', this.selectNode);
    this.contentDOM = document.createElement('span');
    this.dom.appendChild(this.contentDOM);
    if (this.node.attrs.figIds === null) {
      this.open();
    }
  }

  selectNode = (): void => {
    this.dom.classList.add('ProseMirror-selectednode');
    this.open();
  };

  deselectNode(): void {
    this.dom.classList.remove('ProseMirror-selectednode');
    this.close();
  }

  update(node: ProsemirrorNode): boolean {
    this.node = node;
    return true;
  }

  open(): void {
    this.figureEditorContainer = this.view.dom.parentNode.appendChild(document.createElement('div'));

    ReactDOM.render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <FigureCitationEditorPopup
            selectedIds={this.node.attrs.figIds || []}
            anchorEl={this.dom}
            onClose={this.close}
            onChange={this.handleChange}
          />
        </ThemeProvider>
      </Provider>,
      this.figureEditorContainer
    );
  }

  stopEvent(): boolean {
    return true;
  }

  ignoreMutation(): boolean {
    return true;
  }

  close = (): void => {
    this.dom.classList.remove('ProseMirror-selectednode');
    if (this.figureEditorContainer) {
      ReactDOM.unmountComponentAtNode(this.figureEditorContainer);
      this.figureEditorContainer.parentNode.removeChild(this.figureEditorContainer);
      this.figureEditorContainer = null;
    }
  };

  handleChange = (figures: FiguresListEntry[]): void => {
    const ids = figures.map(({ id }) => id);
    const change = this.view.state.tr.setNodeMarkup(this.getPos(), null, { figIds: ids });

    if (this.node.textContent === '???' && figures.length > 0) {
      const nodeContent = (figures[0].name || UNLABELLED_FIGURE_TEXT).replace(/\.$/, '');
      change
        .setSelection(TextSelection.create(change.doc, this.getPos() + 1, this.getPos() + this.node.nodeSize - 1))
        .replaceSelectionWith(this.view.state.schema.text(nodeContent));
    } else if (figures.length === 0) {
      change
        .setSelection(TextSelection.create(change.doc, this.getPos() + 1, this.getPos() + this.node.nodeSize - 1))
        .replaceSelectionWith(this.view.state.schema.text('???'));
    }

    this.view.dispatch(change);
  };
}
