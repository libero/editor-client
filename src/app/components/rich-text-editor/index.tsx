import React from 'react';
import { EditorView } from 'prosemirror-view';
import { EditorState, Transaction, TextSelection } from 'prosemirror-state';
import { debounce, get } from 'lodash';
import { DOMSerializer, Slice } from 'prosemirror-model';
import 'prosemirror-example-setup/style/style.css';
import 'prosemirror-menu/style/menu.css';
import { SectionContainer, SectionContainerVariant } from 'app/components/section-container';
import { ReferenceCitationNodeView } from 'app/components/reference-citation-editor-popup';
import { ComponentWithId } from 'app/utils/types';
import { BoxTextNodeView } from 'app/components/box-text';
import { hasParentNodeOf } from 'app/utils/view.utils';
import { LinkNodeView } from 'app/components/link-editor-popup';
import { FigureNodeView } from 'app/components/figure';
import { FigureCitationNodeView } from 'app/components/figure-citation';

/*
 * available classes:
 *   - root: applied to SectionContainer
 * */

export interface RichTextEditorProps {
  editorState: EditorState;
  label?: string;
  classes?: Record<string, string>;
  name?: string;
  isActive: boolean;
  variant?: SectionContainerVariant;
  onChange?: (change: Transaction, name: string) => void;
  onFocus?: (state: EditorState, name: string) => void;
  onBlur?: (state: EditorState, name: string) => void;
}

const restoreSelection = debounce((editorView, from, to) => {
  if (editorView.state.selection) {
    const $from = editorView.state.doc.resolve(from);
    const $to = editorView.state.doc.resolve(to);
    const change = editorView.state.tr.setSelection(new TextSelection($from, $to));
    editorView.dispatch(change);
  }
}, 50);

export class RichTextEditor extends React.Component<ComponentWithId<RichTextEditorProps>> {
  private options;
  public editorView: EditorView;

  constructor(props: ComponentWithId<RichTextEditorProps>) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
    this.options = {
      nodeViews: {
        refCitation(node, view, getPos) {
          return new ReferenceCitationNodeView(node, view, getPos);
        },
        boxText: (node, view, getPos) => {
          return new BoxTextNodeView(node, view, getPos);
        },
        link(node, view) {
          return new LinkNodeView(node, view);
        },
        figure: (node, view, getPos) => {
          return new FigureNodeView(node, view, getPos);
        },
        figureCitation: (node, view, getPos) => {
          return new FigureCitationNodeView(node, view, getPos);
        }
      },
      handleDOMEvents: {
        focus: ({ state }: EditorView) => {
          if (this.props.onFocus && !this.props.isActive) {
            this.props.onFocus(state, this.props.name);
          }
          return true;
        },
        blur: ({ state }: EditorView) => {
          if (this.props.onBlur && this.props.isActive) {
            this.props.onBlur(state, this.props.name);
          }
          return true;
        }
      }
    };
  }

  focus() {
    this.focusEditor();
  }

  blur() {
    (this.editorView.dom as HTMLDivElement).blur();
  }

  shouldComponentUpdate(nextProps: RichTextEditorProps) {
    // editor state is kept in sync between app state and editor state. App state will change when using formatting
    // or any toolbar menu. In this case when change comes from outside it needs to override editor state
    if (!nextProps.editorState.doc.eq(this.props.editorState.doc)) {
      this.updateEditorState(nextProps.editorState);
    }

    return (
      this.props.label !== nextProps.label ||
      this.props.isActive !== nextProps.isActive ||
      this.props.variant !== nextProps.variant
    );
  }

  componentDidUpdate() {
    // when component updates we need to restore focus
    this.restoreFocus();
  }

  componentWillUnmount() {
    if (this.editorView) {
      this.editorView.destroy();
    }
  }

  updateEditorState(editorState: EditorState) {
    if (editorState !== this.editorView.state) {
      this.editorView.updateState(editorState);
    }
  }

  render() {
    return (
      <SectionContainer
        label={this.props.label}
        className={get(this.props, 'classes.root', '')}
        focused={this.props.isActive}
        id={this.props.id}
        variant={this.props.variant}
      >
        <div ref={this.createEditorView} className="prosemirrorContainer" />
      </SectionContainer>
    );
  }

  private async handleChange(change: Transaction) {
    const newState = this.editorView.state.apply(change);
    this.updateEditorState(newState);
    this.props.onChange(change, this.props.name);
  }

  private createEditorView = (element: HTMLElement) => {
    if (element) {
      const clipboardSerializer = DOMSerializer.fromSchema(this.props.editorState.schema);
      Object.entries(this.props.editorState.schema.nodes).forEach(([nodeName, nodeType]) => {
        clipboardSerializer.nodes[nodeName] = get(nodeType, 'spec.toClipboardDOM', clipboardSerializer.nodes[nodeName]);
      });

      const additionalOptions = this.options || {};
      this.editorView = new EditorView(element, {
        ...additionalOptions,
        clipboardSerializer,
        handleScrollToSelection: (view: EditorView) => {
          return this.isFocusControlDelegated();
        },
        clipboardTextSerializer,
        state: this.props.editorState,
        dispatchTransaction: this.handleChange
      });
    }
  };

  private focusEditor() {
    const { from, to } = this.editorView.state.selection;
    restoreSelection(this.editorView, from, to);
    this.editorView.focus();
  }

  private isFocusControlDelegated() {
    const { $from } = this.editorView.state.selection;
    return this.editorView.dom.contains(document.activeElement) || hasParentNodeOf($from, ['boxText', 'figure']);
  }

  // focus is restored based on 3 conditions
  // #1 - component is active acc. to application state (prop: isActive)
  // #2 - editor does not have focus
  // #3 - focus is not delegated to a node view
  private restoreFocus() {
    if (this.props.isActive && this.editorView && !this.editorView.hasFocus()) {
      if (!this.isFocusControlDelegated()) {
        this.focusEditor();
      }
    }
  }
}

function clipboardTextSerializer(slice: Slice): string {
  const content = slice.content;
  const text = [];
  let separated = true;
  const blockSeparator = ' ';
  content.nodesBetween(0, content.size, (node, pos) => {
    if (node.isText) {
      text.push(node.text);
      separated = !blockSeparator;
    } else if (node.isLeaf && node.type.spec.toClipboardText) {
      text.push(node.type.spec.toClipboardText(node));
      separated = !blockSeparator;
    } else if (!separated && node.isBlock) {
      text.push(blockSeparator);
      separated = true;
    }
    return true;
  });

  return text.join('');
}
