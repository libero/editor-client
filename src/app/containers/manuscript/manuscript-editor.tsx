import React, { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { EditorState, Transaction } from 'prosemirror-state';

import { getAbstract, getKeywordGroups, getTitle } from 'app/selectors/manuscript.selectors';
import * as manuscriptActions from 'app/actions/manuscript.actions';
import { RichTextEditor } from 'app/components/rich-text-editor';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { KeywordsEditor } from 'app/components/keywords';
import { KeywordGroups } from 'app/models/manuscript';
import { useManuscriptStyles } from './styles';
import { SortableAuthorsList } from './sortable-authors-list';
import { AuthorsInfoDetails } from 'app/containers/manuscript/authors-info-details';
import { AffiliationsList } from 'app/containers/manuscript/affiliations-list';
import { ReferenceList } from 'app/containers/manuscript/references-list';
import { ArticleInformation } from 'app/containers/manuscript/article-information';
import { getFocusedEditorStatePath } from 'app/selectors/manuscript-editor.selectors';

const isInputFocused = (inputName: string, focusedPath?: string) => {
  return Boolean(focusedPath) && focusedPath.startsWith(inputName);
};

export const ManuscriptEditor: React.FC = () => {
  const classes = useManuscriptStyles();
  const dispatch = useDispatch();

  const title: EditorState = useSelector(getTitle);
  const abstract: EditorState = useSelector(getAbstract);
  const focusedPath = useSelector(getFocusedEditorStatePath);
  const allKeywords: KeywordGroups = useSelector(getKeywordGroups);

  const handleTitleChange = useCallback(
    (diff: Transaction) => {
      dispatch(manuscriptActions.updateTitleAction(diff));
    },
    [dispatch]
  );

  const handleKeywordsChange = (keywordGroup: string, index: number, diff: Transaction): void => {
    dispatch(manuscriptActions.updateKeywordAction({ keywordGroup, index, change: diff }));
  };

  const handleNewKeywordChange = (keywordGroup: string, diff: Transaction): void => {
    dispatch(manuscriptActions.updateNewKeywordAction({ keywordGroup, change: diff }));
  };

  const handleAbstractChange = useCallback(
    (diff: Transaction): void => {
      dispatch(manuscriptActions.updateAbstractAction(diff));
    },
    [dispatch]
  );

  const handleKeywordDelete = (keywordGroup: string, index: number): void => {
    dispatch(manuscriptActions.deleteKeywordAction({ keywordGroup, index }));
  };

  const handleKeywordAdd = (keywordGroup: string, keyword: EditorState): void => {
    dispatch(manuscriptActions.addNewKeywordAction({ keywordGroup, keyword }));
  };

  const handleFocusSwitch = useCallback(
    (state: EditorState, path: string) => {
      dispatch(manuscriptEditorActions.setFocusAction(path));
    },
    [dispatch]
  );

  const handleBlur = useCallback(() => {
    dispatch(manuscriptEditorActions.removeFocusAction());
  }, [dispatch]);

  const handleKeywordFocus = (group: string, index: number | undefined, isNewKeywordFocused: boolean): void => {
    const kwdIndexPath = isNewKeywordFocused ? 'newKeyword' : `keywords.${index}`;
    handleFocusSwitch(null, ['keywordGroups', group, kwdIndexPath].join('.'));
  };

  const renderKeywords = (keywordGroups: KeywordGroups): JSX.Element[] => {
    return Object.entries(keywordGroups).map(([groupType, group]) => {
      return (
        <KeywordsEditor
          key={groupType}
          keywords={group.keywords}
          newKeyword={group.newKeyword}
          activeKeywordPath={focusedPath}
          name={groupType}
          label={group.title || 'Keywords'}
          onNewKeywordChange={handleNewKeywordChange}
          onAdd={handleKeywordAdd}
          onChange={handleKeywordsChange}
          onDelete={handleKeywordDelete}
          onFocusSwitch={handleKeywordFocus}
          onBlur={handleBlur}
        />
      );
    });
  };

  return (
    <div className={classes.content}>
      <div aria-hidden="true" className={classes.toolbarPlaceholder} />
      <RichTextEditor
        editorState={title}
        label="Title"
        isActive={isInputFocused('title', focusedPath)}
        name="title"
        onChange={handleTitleChange}
        onFocusSwitch={handleFocusSwitch}
      />
      <SortableAuthorsList />
      <AffiliationsList />
      <RichTextEditor
        editorState={abstract}
        label="Abstract"
        name="abstract"
        isActive={isInputFocused('abstract', focusedPath)}
        onChange={handleAbstractChange}
        onFocusSwitch={handleFocusSwitch}
      />
      {renderKeywords(allKeywords)}
      <ArticleInformation />
      <AuthorsInfoDetails />
      <ReferenceList />
    </div>
  );
};
