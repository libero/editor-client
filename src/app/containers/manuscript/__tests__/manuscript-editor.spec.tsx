import React from "react";
import { create } from "react-test-renderer";
import configureMockStore from 'redux-mock-store'
import {Provider} from 'react-redux'

import {ManuscriptEditor} from "../manuscript-editor";
import {
  getInitialLoadableState,
  getLoadableStateSuccess
} from "../../../utils/state.utils";
import {RichTextEditor} from "../../../components/rich-text-editor";

describe('manuscript editor', () => {
  const mockStore = configureMockStore([]);

  it('should render', () => {
    const store = mockStore({ manuscript: getInitialLoadableState() });
    const wrapper = create(<Provider store={store}>
      <ManuscriptEditor />
    </Provider>);

    expect(wrapper).toMatchSnapshot();
  });

  it('should update state when RichTextEditor changes', () => {
    const mockEditorState = {
      title: { apply: jest.fn() }
    };

    const store = mockStore({ manuscript: getLoadableStateSuccess(mockEditorState) });
    const wrapper = create(<Provider store={store}> <ManuscriptEditor /> </Provider>);
    const changeArg = Symbol();
    wrapper.root.findByType(RichTextEditor).props.onChange(changeArg);

    expect(mockEditorState.title.apply).toBeCalledWith(changeArg);
  });
})