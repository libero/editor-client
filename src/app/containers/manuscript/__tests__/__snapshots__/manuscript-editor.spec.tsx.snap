// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manuscript editor should render 1`] = `
<div
  className="makeStyles-contentWrapper-1"
  data-test-id="container-wrapper"
  onClick={[Function]}
>
  <div
    aria-hidden="true"
    className="makeStyles-toolbarPlaceholder-4"
  />
  <div
    className="makeStyles-content-2"
    onClick={[Function]}
  >
    <div
      className="makeStyles-root-5 makeStyles-plainVariant-7"
    >
      <label
        className="makeStyles-label-8"
      >
        Title
      </label>
      <div
        className="prosemirrorContainer"
      />
    </div>
    <div
      onClick={[Function]}
    >
      <div
        aria-hidden="true"
        className="makeStyles-spacer-3"
        onClick={[Function]}
      />
      <section>
        <div
          className="makeStyles-root-5 makeStyles-plainVariant-7"
        >
          <label
            className="makeStyles-label-8"
          >
            Authors
          </label>
          <div
            className="makeStyles-sortableContainer-10"
          >
             
             
          </div>
        </div>
        <button
          className="MuiButtonBase-root MuiButton-root makeStyles-addEntityButton_root-17 MuiButton-text makeStyles-addAuthorButton-13 MuiButton-textSizeSmall MuiButton-sizeSmall MuiButton-disableElevation"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            <span
              className="MuiButton-startIcon makeStyles-addEntityButton_icon-18 MuiButton-iconSizeSmall"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
            </span>
            Author
          </span>
        </button>
      </section>
      <div
        aria-hidden="true"
        className="makeStyles-spacer-3"
        onClick={[Function]}
      />
      <section>
        <div
          className="makeStyles-root-5 makeStyles-plainVariant-7"
        >
          <label
            className="makeStyles-label-8"
          >
            Affiliations
          </label>
        </div>
        <button
          className="MuiButtonBase-root MuiButton-root makeStyles-addEntityButton_root-17 MuiButton-text makeStyles-addButton-19 MuiButton-textSizeSmall MuiButton-sizeSmall MuiButton-disableElevation"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            <span
              className="MuiButton-startIcon makeStyles-addEntityButton_icon-18 MuiButton-iconSizeSmall"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
            </span>
            Affiliation
          </span>
        </button>
      </section>
      <div
        aria-hidden="true"
        className="makeStyles-spacer-3"
        onClick={[Function]}
      />
    </div>
    <div
      className="makeStyles-root-5 makeStyles-plainVariant-7"
    >
      <label
        className="makeStyles-label-8"
      >
        Abstract
      </label>
      <div
        className="prosemirrorContainer"
      />
    </div>
    <div
      aria-hidden="true"
      className="makeStyles-spacer-3"
      onClick={[Function]}
    />
    <div
      className="makeStyles-root-5 makeStyles-plainVariant-7"
    >
      <label
        className="makeStyles-label-8"
      >
        Impact statement
      </label>
      <div
        className="prosemirrorContainer"
      />
    </div>
    <div
      aria-hidden="true"
      className="makeStyles-spacer-3"
      onClick={[Function]}
    />
    <div
      onClick={[Function]}
    >
      <div
        aria-hidden="true"
        className="makeStyles-spacer-3"
        onClick={[Function]}
      />
      <div
        className="makeStyles-sectionContainer-25 makeStyles-root-5 makeStyles-plainVariant-7"
      >
        <label
          className="makeStyles-label-8"
        >
          Article information
        </label>
        <section
          className="makeStyles-infoSection-26"
        >
          <div>
            <strong>
               Article DOI: 
            </strong>
             
            
          </div>
        </section>
        <button
          className="MuiButtonBase-root MuiIconButton-root makeStyles-editButton-27"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        aria-hidden="true"
        className="makeStyles-spacer-3"
        onClick={[Function]}
      />
      <section>
        <div
          className="makeStyles-root-5 makeStyles-plainVariant-7"
        >
          <label
            className="makeStyles-label-8"
          >
            Related articles
          </label>
          <ul
            className="makeStyles-list-29"
          />
        </div>
        <button
          className="MuiButtonBase-root MuiButton-root makeStyles-addEntityButton_root-17 MuiButton-text MuiButton-textSizeSmall MuiButton-sizeSmall MuiButton-disableElevation"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            <span
              className="MuiButton-startIcon makeStyles-addEntityButton_icon-18 MuiButton-iconSizeSmall"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
            </span>
            Article
          </span>
        </button>
      </section>
      <div
        aria-hidden="true"
        className="makeStyles-spacer-3"
        onClick={[Function]}
      />
      <section>
        <div
          className="makeStyles-root-5 makeStyles-plainVariant-7"
        >
          <label
            className="makeStyles-label-8"
          >
            Author Information
          </label>
        </div>
        <button
          className="MuiButtonBase-root MuiButton-root makeStyles-addEntityButton_root-17 MuiButton-text makeStyles-addAuthorButton-33 MuiButton-textSizeSmall MuiButton-sizeSmall MuiButton-disableElevation"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            <span
              className="MuiButton-startIcon makeStyles-addEntityButton_icon-18 MuiButton-iconSizeSmall"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
            </span>
            Author
          </span>
        </button>
      </section>
      <div
        aria-hidden="true"
        className="makeStyles-spacer-3"
        onClick={[Function]}
      />
      <section>
        <div
          className="makeStyles-root-5 makeStyles-plainVariant-7"
        >
          <label
            className="makeStyles-label-8"
          >
            References
          </label>
          <ul
            className="makeStyles-list-34"
          />
        </div>
        <button
          className="MuiButtonBase-root MuiButton-root makeStyles-addEntityButton_root-17 MuiButton-text MuiButton-textSizeSmall MuiButton-sizeSmall MuiButton-disableElevation"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            <span
              className="MuiButton-startIcon makeStyles-addEntityButton_icon-18 MuiButton-iconSizeSmall"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
            </span>
            Reference
          </span>
        </button>
      </section>
    </div>
  </div>
</div>
`;
